hubtools/src/lib.rs

```diff
@@ -666,9 +683,24 @@ impl RawHubrisArchive {
             .as_str()
             .ok_or(Error::BadTomlType)?
             .to_owned();
+        // Hubris has an epoch in the app.toml but Bootleby does not.
+        // CLI arg overrides app.toml, value will default to zero.
+        let epoc: u32 = if let Some(param) = epoch {
+            param.parse::<u32>().map_err(|_| Error::InvalidEpoch(param.to_string()))?
+        } else if let Some(param) = manifest.as_table().ok_or(Error::BadTomlType)?.get("epoch") {
+            let toml_int = param.as_integer().ok_or(Error::BadTomlType)?;
+            u32::try_from(toml_int).map_err(|_| Error::InvalidEpoch(format!("epoch from toml = {}", toml_int)))?
+        } else {
+            // Default is zero if missing from app.toml
+            0u32
+        };
+        // EPOC in the caboose is a u32, zero-padded 10-digit ascii number.
+        // Missing EPOC or all-zeros is equivalent to EPOC zero.
+        // Anything else represents some future scheme or a typo in toml file.
+        let caboose_epoc = format!("{:010}", epoc).to_owned();
```

Comment:
The epoch was added in the app.toml as a random thought. Is there value in keeping it there vs getting surprised later?

I'm also not a huge fan of the parsing then re-formatting. I'd rather require the user to specify the epoch exactly and then reject it if it isn't in the expected form

Reply:
This can be put into methods on an Epoch struct.
We have the line "epoch = 0" in app.toml files which Toml returns as an i64.
The CLI should mirror that but clap could return an i64 as well instead of a String.
I could remove the leading ASCII zeros. I had thought originally to compare EPOC byte by byte, but I wound up converting to u32 to stay consistent with ImageHeader's representation.

Reply:
Re: "The epoch was added in app.toml as a random thought..." 

The ImageHeader `version` isn't expected to be used, but the `epoch` field has a very nice property for rollback protection: it is available to `update_server` before any flash is erased.

This allows us to reject updates early and keeps us from wearing out flash if some person or process keeps retrying because they don't understand that rollback protection is keeping the update from being applied.

Reply:
We should think about our release flow. We manage version strings manually and epoch could be the same.
That is very helpful for SP hubris which is compiled differently for the different target boards. We might have some dangerous PSC SP release that we  want to ban with rollback protection. We would increment epoch there but wouldn't have to for the other platforms. Using the app.toml  to track epoch works there because of the different app.toml files per target.

We don't have an app.toml for bootleby. I would like to find a git-controlled place to put it though.

Reply:
Managing the version string for releases needs more discussion.
We could put a json file in sprot-release that binds (repo, commit, version) or (repo, commit, version, epoch).
But that source of truth is redundant with artifacts that permslip stores. We could call it a release plan.
We could just store the version to be used.
We could aromatically query production permslip on the next available major or minor version.

Reply:
Added optional `--epoch u32` to `--package-elf` as you suggested.